### PR TITLE
Hold resource until stream completes.

### PR DESCRIPTION
**Pull request by isoos**
It needs to track when the underlying stream completes, and it shall propagate cancels. To fulfill both requirements, neither StreamCompleter, nor StreamController.addStream is sufficient.

### CODE REVIEW

**Comment by isoos** [comment]
I signed it!

File: lib/pool.dart [hunk 1]
```diff
@@ -121,6 +121,44 @@ class Pool {
...
+      if (resourceCompleter.isCompleted) return null;
+      new Future.sync(callback).then((Stream/*<T>*/ stream) {
+        if (stream == null) {
+          complete();
+          return null;
+        }
+        streamSubscription = stream.listen(streamController.add,
+            onError: streamController.addError,
+            onDone: complete,
+            cancelOnError: true);
```
Comment (nex3) [inline lib/pool.dart]: Rather than manually forwarding events, use [`new StreamTransformer.fromHandlers()`](https://api.dartlang.org/stable/1.21.0/dart-async/StreamTransformer/StreamTransformer.fromHandlers.html) to release the resource on a "done" event. That'll ensure that the resulting stream is *cancel-correct*, meaning that it properly forwards cancellations and pauses to the underlying stream.

File: lib/pool.dart [hunk 2]
```diff
@@ -121,6 +121,44 @@ class Pool {
...
+        streamSubscription.cancel();
+        streamSubscription = null;
+      }
+      if (!resourceCompleter.isCompleted) {
+        resourceCompleter.complete();
+        streamController.close();
+      }
+    };
+    streamController = new StreamController/*<T>*/(onCancel: complete);
+    withResource(() {
```
Comment (nex3) [inline lib/pool.dart]: I think it'll be cleaner to manually allocate a resource using `request()`, since you're releasing it in a stream callback anyway.

File: lib/pool.dart [hunk 3]
```diff
@@ -121,6 +121,44 @@ class Pool {
...
+      if (resourceCompleter.isCompleted) return null;
+      new Future.sync(callback).then((Stream/*<T>*/ stream) {
```
Comment (nex3) [inline lib/pool.dart]: I don't think we should run `callback()` in `new Future.sync()`. The type annotation says it can only return a `Stream`, not `Future<Stream>`.

File: test/pool_test.dart [hunk 4]
```diff
@@ -105,6 +105,68 @@ void main() {
     });
   });
 
+  group("streamWithResource()", () {
+    test("The Stream holds a resource until it is completed.", () {
+      new FakeAsync().run((async) {
+        StreamController<int> streamController = new StreamController<int>();
```
Comment (nex3) [inline test/pool_test.dart]: Nit: Use "var" for local variables.

File: test/pool_test.dart [hunk 5]
```diff
@@ -105,6 +105,68 @@ void main() {
...
+        bool nextTaskCompleted = false;
+        pool.withResource(() {
+          nextTaskCompleted = true;
+        });
+
+        new Future.delayed(new Duration(microseconds: 1), () {
+          expect(nextTaskCompleted, isFalse);
+          streamController.close();
+          return new Future.delayed(new Duration(microseconds: 1));
+        }).then((_) {
```
Comment (nex3) [inline test/pool_test.dart]: The other tests were written before `async`/`await` existed, but new tests should use them in preference to `.then()`.

File: test/pool_test.dart [hunk 6]
```diff
@@ -105,6 +105,68 @@ void main() {
...
+      });
+    });
+
+    test("Cancelling the Stream propagates and releases the resource.", () {
+      new FakeAsync().run((async) {
+        StreamController<int> streamController = new StreamController<int>();
+        var pool = new Pool(1);
+        StreamSubscription<int> subscription = pool
+            .streamWithResource(() => streamController.stream)
+            .listen((_) {});
```
Comment (nex3) [inline test/pool_test.dart]: Fun fact: you can write `listen(null)` if you don't want to do anything with the data events.

File: test/pool_test.dart [hunk 7]
```diff
@@ -105,6 +105,68 @@ void main() {
...
+        pool.withResource(() {
+          nextTaskCompleted = true;
+        });
+
+        new Future.delayed(new Duration(microseconds: 1), () {
+          expect(nextTaskCompleted, isFalse);
+          subscription.cancel();
+          return new Future.delayed(new Duration(microseconds: 1));
+        }).then((_) {
+          expect(nextTaskCompleted, isTrue);
```
Comment (nex3) [inline test/pool_test.dart]: Also assert that the original `StreamController` received a cancel event.

File: test/pool_test.dart [hunk 8]
```diff
@@ -105,6 +105,68 @@ void main() {
...
+        });
+
+        async.elapse(new Duration(seconds: 1));
+      });
+    });
+
+    test("Simple Stream example.", () {
+      new FakeAsync().run((async) {
```
Comment (nex3) [inline test/pool_test.dart]: I don't think you need `FakeAsync` for this test. You can just do `expect(pool.streamWithResource(...).toList(), completion(equals([0, 1, 2])))`.

File: test/pool_test.dart [hunk 9]
```diff
@@ -105,6 +105,68 @@ void main() {
...
+        var pool = new Pool(1);
+        pool.streamWithResource(() => new Stream.fromIterable([0, 1, 2]))
+            .toList()
+            .then((list) {
+              expect(list, [0, 1, 2]);
+            });
+        async.elapse(new Duration(seconds: 1));
+      });
+    });
```
Comment (nex3) [inline test/pool_test.dart]: Also test that the resource is released if the callback returns `null`, and that the callback isn't run until a resource is available.

File: lib/pool.dart [hunk 10]
```diff
@@ -121,6 +121,44 @@ class Pool {
...
+    final complete = () {
+      if (streamSubscription != null) {
+        streamSubscription.cancel();
+        streamSubscription = null;
+      }
+      if (!resourceCompleter.isCompleted) {
+        resourceCompleter.complete();
+        streamController.close();
+      }
+    };
```
Comment (isoos) [inline lib/pool.dart]: I've tried that, but either the streams' complete or the cancel event does not propagate (don't remember which one).

File: lib/pool.dart (hunk 2 above)
Comment (isoos) [inline lib/pool.dart]: I think we need similar error handling as we have around the `withResource()`, and using `request()` will just duplicate it (adding a few more lines of code). E.g. the `resource.release()` should be placed in `complete` to get it consistent with the stream handler -> separate field definition, value allocation etc.
If you think it is worth it, I would do it, but I like it this way better.

Comment (nex3) [inline lib/pool.dart]: I don't think the cleanup logic is that onerous—just a `try`/`finally`. If you think you can get it cleaner with `withResource()`, though, go ahead, but I may ask you to switch it to `request()` later.

File: lib/pool.dart (hunk 1 above)
Comment (isoos) [inline lib/pool.dart]: I've addressed most of the comments, but I'm stuck with this. I was not able to use `StreamTransformer` (w/ `StreamCompleter`) to get the tests passing. What kind of code would you like to see?

Comment (nex3) [inline lib/pool.dart]: Just something like:

```dart
return stream.transform(new StreamTransformer.fromHandlers(handleDone: (sink) {
  resource.release();
  sink.close();
});  
```

Comment (isoos) [inline lib/pool.dart]: But then the signature of the method becomes `Future<Stream<T>>`. While both provide value, I think the library should support the harder one, which is only the `Stream<T>`.

Comment (nex3) [inline lib/pool.dart]: That's what the `StreamCompleter` is for—it converts a `Future<Stream<T>>` to a `Stream<T>`.

Comment (isoos) [inline lib/pool.dart]: I've got this version, but 3 of the tests don't pass with it. Any idea?

```
  Stream/*<T>*/ streamWithResource/*<T>*/(Stream/*<T>*/ callback()) {
    StreamCompleter/*<T>*/ streamCompleter = new StreamCompleter/*<T>*/();
    PoolResource poolResource;
    request().then((resource) {
      poolResource = resource;
      try {
        Stream/*<T>*/ stream = callback();
        if (stream == null) {
          streamCompleter.setEmpty();
          poolResource.release();
          return null;
        }
        streamCompleter.setSourceStream(stream.transform(
            new StreamTransformer.fromHandlers(handleDone: (sink) {
              sink.close();
              poolResource.release();
            })));
      } catch (e, st) {
        streamCompleter.setError(e, st);
        poolResource.release();
      }
    }, onError: streamCompleter.setError);
    return streamCompleter.stream;
  }
```

Comment (nex3) [inline lib/pool.dart]: Can you give me some more details? Which tests don't pass? How are they failing?

Comment (isoos) [inline lib/pool.dart]: I've fixed one with an extra `await new Future.delayed(new Duration());`, but two tests are still failing:
- The Stream holds a resource until it is completed.
- Cancelling the Stream propagates and releases the resource.

In both cases it fails on the last line: asserting that the next task will get completed too. (`expect(nextTaskCompleted, isTrue);`)

Comment (nex3) [inline lib/pool.dart]: It looks like the updated tests don't use `FakeAsync` anymore. Why is that? Could the failures be caused by a race condition?

If you update the PR with the failing tests, I can patch it in and see what I can figure out.

Comment (isoos) [inline lib/pool.dart]: Added it to a new branch: https://github.com/dart-lang/pool/compare/master...isoos:stream2?expand=1

I don't think async/await vs. FakeAsync made a difference, I had issues with StreamCompleter in the very early version too. (see initial comment on the PR).

Comment (isoos) [inline lib/pool.dart]: I've also realized that I will need an easy-to-access on-completion signal to release the database connection. I've done a prototype for that here:
https://github.com/dart-lang/pool/compare/master...isoos:stream3?expand=1

Comment (nex3) [inline lib/pool.dart]: > Added it to a new branch: https://github.com/dart-lang/pool/compare/master...isoos:stream2?expand=1

For "the stream holds a resource until it is completed", the test isn't listening to the stream, which means the done event is never getting emitted. If you add `.listen(null)` it works.

The cancel test is trickier. If a stream is canceled, it'll never actually emit a done event, but there's not currently a great way to intercept cancellations gracefully. That's probably a hole in our APIs; if you're willing to wait a day or two, I can work on a patch to the `async` package that'll add a transformer that does that nicely.

> I've also realized that I will need an easy-to-access on-completion signal to release the database connection. I've done a prototype for that here:
> https://github.com/dart-lang/pool/compare/master...isoos:stream3?expand=1

Can't you just use the returned stream's "done" event to check this? Using a callback (or a future-back) feels like a violation of the layering philosophy of this package. We generally expect the user to tell pool when resources are being used or when they're not, and this inverts that.

Comment (isoos) [inline lib/pool.dart]: > If you add .listen(null) it works.

Not sure how this should work. The returned Future of `withResource` will complete, it doesn't matter if we add an `await` or `.then()` on it. I think the same should be true for Streams (e.g. it should complete with or without a subscriber), but can be convinced otherwise.

> if you're willing to wait a day or two

sure, thanks for jumping on this!

> Can't you just use the returned stream's "done" event to check this?

How can I listen on Stream's done? The `drain` will introduce another listener, which will fail if it is single-subscription. `StreamController.done` is not available when all I got is a Stream.

Comment (nex3) [inline lib/pool.dart]: > Not sure how this should work. The returned Future of `withResource` will complete, it doesn't matter if we add an `await` or `.then()` on it. I think the same should be true for Streams (e.g. it should complete with or without a subscriber), but can be convinced otherwise.

This is one of the differences between Futures and Streams in general: Futures don't care about whether a listener is registered and streams do (or at least, they can). Given that, I think it makes sense for the behavior to differ here. In practice, I don't expect anyone will realistically be passing streams through only to discard them later. It's more likely that they'll want to make use of the laziness and memory benefits that come from following stream conventions here.

> sure, thanks for jumping on this!

I'll cc you on the pull request.

> How can I listen on Stream's done? The `drain` will introduce another listener, which will fail if it is single-subscription. `StreamController.done` is not available when all I got is a Stream.

You can always use a transformer to hook into a stream's events without consuming it, just like you're doing in this PR already.

File: lib/pool.dart [hunk 11]
```diff
@@ -121,6 +121,55 @@ class Pool {
     });
   }
 
+  /// Requests a resource for the combined duration of the [callback] and the
+  /// processing of the Stream it returns.
+  /// (E.g. a database connection is hold until the the results get processed.)
+  Stream/*<T>*/ streamWithResource/*<T>*/(Stream/*<T>*/ callback()) {
```
Comment (kevmoo) [inline lib/pool.dart]: Use the "real" generic method syntax. `streamWithResource<T>` etc

Here and below

**Comment by kevmoo** [comment]
what's the story here @isoos @nex3 ?

**Comment by isoos** [comment]
> what's the story here @isoos @nex3 ?

I think this became obsolete in the past few years. I'm also wondering if the same functionality could be written with a simpler structure e.g. using `async*` in the main method.
